### PR TITLE
Joao/add invoice entity

### PR DESCRIPTION
**Pull request by devjoaov**
## Pull Request

**Summary:**

**Related Issue:**

**Description:**

**Testing:**

**Screenshots (if applicable):**

**Checklist:**

- [ ] Code follows the project's coding standards
- [ ] Tests have been added or updated to cover the changes
- [ ] Documentation has been updated (if applicable)
- [ ] All existing tests pass
- [ ] Conforms to the contributing guidelines

**Additional Notes:**

### CODE REVIEW

File: lib/pay/stripe.rb [hunk 1]
```diff
@@ -117,6 +123,29 @@ def self.configure_webhooks
...
+        events.subscribe "stripe.invoice.payment_failed", Pay::Stripe::Webhooks::PaymentFailed.new
+
+        events.subscribe "stripe.invoice.created", Pay::Stripe::Webhooks::InvoiceCreated.new
+        events.subscribe "stripe.invoice.deleted", Pay::Stripe::Webhooks::InvoiceDeleted.new
+        events.subscribe "stripe.invoice.finalization_failed", Pay::Stripe::Webhooks::InvoiceFinalizationFailed.new
+        events.subscribe "stripe.invoice.finalized", Pay::Stripe::Webhooks::InvoiceFinalized.new
+        events.subscribe "stripe.invoice.overdue", Pay::Stripe::Webhooks::InvoiceOverdue.new
+        events.subscribe "stripe.invoice.paid", Pay::Stripe::Webhooks::InvoicePaid.new
+        events.subscribe "stripe.invoice.payment_succeeded", Pay::Stripe::Webhooks::InvoicePaymentSucceeded.new
+        # UPCOMING HAS NO ID
```
Comment (ribeirojose) [inline lib/pay/stripe.rb:140]: really? 👀 when is this fired?

File: lib/pay/stripe.rb [hunk 2]
```diff
@@ -117,6 +123,29 @@ def self.configure_webhooks
...
+        # events.subscribe "stripe.invoice.upcoming", Pay::Stripe::Webhooks::InvoiceUpcoming.new
+        events.subscribe "stripe.invoice.updated", Pay::Stripe::Webhooks::InvoiceUpdated.new
+        events.subscribe "stripe.invoice.voided", Pay::Stripe::Webhooks::InvoiceVoided.new
+        events.subscribe "stripe.invoice.will_be_due", Pay::Stripe::Webhooks::InvoiceWillBeDue.new
+        # events.subscribe "stripe.invoice.marked_uncollectible", 
+        # events.subscribe "stripe.invoice.payment_action_required", 
+        # events.subscribe "stripe.invoice.payment_failed",
```
Comment (ribeirojose) [inline lib/pay/stripe.rb:147]: when does this fire?

Comment (ribeirojose) [inline lib/pay/stripe.rb:147]: these are repeated? L129-131, right?

File: db/migrate/2_create_pay_invoices.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      t.belongs_to :subscription, foreign_key: {to_table: :pay_subscriptions}, null: true, type: foreign_key_type
+      t.string :processor_id, null: false
+      t.string :status, null: false
+      t.integer :amount_due
+      t.integer :number
+      t.datetime :due_date
+      t.datetime :paid_at
+      t.integer :total
+      t.datetime :period_start
+      t.datetime :period_end
```
Comment (ribeirojose) [inline db/migrate/2_create_pay_invoices.rb:16]: 👌

File: app/models/pay/invoice.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    store_accessor :data, :discounts # array of discount IDs applied to the invoice
+    store_accessor :data, :total_discount_amounts # array of discount details
+    store_accessor :data, :total_tax_amounts # array of tax details for each jurisdiction
+    store_accessor :data, :credit_notes # array of credit notes for the invoice
+    store_accessor :data, :refunds # array of refunds
+
+    delegate_missing_to :pay_processor
+
+    # Helpers for payment processors
+    %w[braintree stripe paddle_billing paddle_classic fake_processor].each do |processor_name|
```
Comment (ribeirojose) [inline app/models/pay/invoice.rb:34]: initially this would only support Stripe, right?

Comment (ribeirojose) [reply app/models/pay/invoice.rb:34]: random question: would this serve us in Asaas as well?

File: lib/pay/stripe/invoice.rb [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        else
+          raise
+        end
+      end
+
+      def self.expand_options
+        {
+          expand: [
+            "customer",
+            "lines",
```
Comment (ribeirojose) [inline lib/pay/stripe/invoice.rb:81]: is this line items?

File: lib/pay/stripe/invoice.rb [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            "charge",
```
Comment (ribeirojose) [inline lib/pay/stripe/invoice.rb:82]: just to reiterate: the charge is created when the invoice is paid, right? So we'd only have it here when the invoice is paid, right? Should charges also belong to an invoice then?

**Review by ribeirojose — APPROVED** [review]
lgtm, left a few questions

**Comment by ribeirojose** [comment]
more generally: should we integrate the Asaas stuff we've built in our app to the lib here as well?

File: lib/pay/stripe/invoice.rb (hunk 6 above)
Comment (devjoaov) [reply lib/pay/stripe/invoice.rb:82]: Yep, we can do this relationship

File: lib/pay/stripe.rb (hunk 1 above)
Comment (devjoaov) [reply lib/pay/stripe.rb:140]: https://docs.stripe.com/billing/subscriptions/webhooks

`Sent a few days prior to the renewal of the subscription. The number of days is based on the number set for Upcoming renewal events in the [Dashboard](https://dashboard.stripe.com/settings/billing/automatic). For existing subscriptions, changing the number of days takes effect on the next billing period. You can still add [extra invoice items](https://docs.stripe.com/billing/invoices/subscription#adding-upcoming-invoice-items), if needed."`

**Pull request merged**